lib/src/models/transaction_uploader.dart

```diff
@@ -183,41 +129,43 @@ class TransactionUploader {
     txJson.remove('data');
     final res = await _api.post('tx', body: json.encode(txJson));
 
-    _lastRequestTimeEnd = DateTime.now().millisecondsSinceEpoch;
-    lastResponseStatus = res.statusCode;
-
     if (!(res.statusCode >= 200 && res.statusCode < 300)) {
       throw StateError('Unable to upload transaction: ${res.statusCode}');
     }
 
     _txPosted = true;
   }
 
-  static Future<TransactionUploader> deserialize(
-      Map<String, dynamic> json, Uint8List data, ArweaveApi api) async {
-    final transaction = json['transaction'] != null
-        ? Transaction.fromJson(json['transaction'])
-        : null;
-
-    await transaction!.setData(data);
-
-    return TransactionUploader._(
-      api: api,
-      chunkIndex: json['chunkIndex'] as int,
-      txPosted: json['txPosted'],
-      transaction: transaction,
-      lastRequestTimeEnd: json['lastRequestTimeEnd'],
-      lastResponseStatus: json['lastResponseStatus'],
-      lastResponseError: json['lastResponseError'],
-    );
-  }
+  /// Uploads the specified chunk onto Arweave.
+  ///
+  /// Throws a [StateError] if the chunk being uploaded encounters a fatal error
+  /// during upload.
+  Future<void> _uploadChunk(int chunkIndex) async {
+    final chunk = _transaction.getChunk(chunkIndex);
+
+    final chunkValid = await validatePath(
+        _transaction.chunks!.dataRoot,
+        int.parse(chunk.offset),
+        0,
+        int.parse(chunk.dataSize),
+        decodeBase64ToBytes(chunk.dataPath));
+
+    if (!chunkValid) {
+      throw StateError('Unable to validate chunk: $chunkIndex');
+    }
 
-  Map<String, dynamic> serialize() => <String, dynamic>{
-        'chunkIndex': _chunkIndex,
-        'txPosted': _txPosted,
-        'transaction': _transaction.toJson()..['data'] = null,
-        'lastRequestTimeEnd': _lastRequestTimeEnd,
-        'lastResponseStatus': lastResponseStatus,
-        'lastResponseError': lastResponseError,
-      };
+    final res = await _api.post('chunk', body: json.encode(chunk));
+
+    if (res.statusCode != 200) {
+      final responseError = getResponseError(res);
+
+      if (fatalChunkUploadErrors.contains(responseError)) {
+        throw StateError(
+            'Fatal error uploading chunk: $chunkIndex: $responseError');
+      } else {
+        throw Exception(
+            'Received non-fatal error while uploading chunk $chunkIndex: $responseError');
+      }
+    }
```

Comment:
One of the gateway errors we've intermittently been experiencing is a 504 with no response body. In these cases there is no statusCode available to check and this would probably result in a type error of some sorts.

It looks like this situation would be fine here during the uploadChunks, as it seems the above method `uploadChunkAndNotifyOfCompletion` would retry the chunk. But the `_postTransactionHeader` method doesn't offer the same protection.

Erroring out on the postTx call isn't that catastrophic because the gateway will never receive the headers and subsequently never charge the wallet. But we could certainly extend the retry to make it more robust.

Reply:
Updated to retry failing header uploads.